Wikipedia/UI-V5/WMFSaveButtonController.m

```diff
@@ -0,0 +1,83 @@
+//  Created by Monte Hurd on 8/19/15.2//  Copyright (c) 2015 Wikimedia Foundation. Provided under MIT-style license; please copy and modify!
+
+#import "WMFSaveButtonController.h"
+#import "MWKSavedPageList.h"
+#import "MWKArticle.h"
+#import "SessionSingleton.h"
+#import "MWKUserDataStore.h"
+
+@interface WMFSaveButtonController ()
+
+@property (strong, nonatomic) UIButton* button;
+
+@end
+
+@implementation WMFSaveButtonController
+
+- (instancetype)initWithButton:(UIButton*)button
+                         title:(MWKTitle*)title {
+    self = [super init];
+    if (self) {
+        self.button = button;
+        self.title  = title;
+        [self observeSavedPages];
+        [self updateSavedButtonState];
+    }
+    return self;
+}
+
+- (MWKSavedPageList*)savedPageList {
+    return [SessionSingleton sharedInstance].userDataStore.savedPageList;
+}
+
+- (void)observeSavedPages {
+    [self.KVOControllerNonRetaining observe:self.savedPageList
+                                    keyPath:WMF_SAFE_KEYPATH(self.savedPageList, entries)
+                                    options:0
+                                      block:^(WMFSaveButtonController* observer, id object, NSDictionary* change) {
+        [observer updateSavedButtonState];
```

Comment:
maybe a premature optimization, but it might save us some time if we checked the object that was added/removed to the list to see if matches our title, e.g.:

``` objc
switch ([changeDictionary[NSKeyValueChangeKindKey] intValue]) {
  case NSKeyValueChangeRemoval:
  case NSKeyValueChangeReplacement: {
    MWKSavedPageEntry* prevEntry = change[NSKeyValueChangeOldKey]
    if (prevEntry.title isEqual:self.title) { /* update state w/o checking entire list */ }
  }
}
```